src/site/cpp-objects.rst

```diff
@@ -0,0 +1,177 @@
+**************************
+C++ Lessons #4 --- Objects
+**************************
+
+.. warning::
+
+    * For these, I will only go over the C++ details at a high level
+    * Getting into the nitty gritty of C++ is not the purpose of this aside
+    * Some of the concepts require knowledge of more advanced ideas that have not been covered yet
+        * If this happens, do not panic
+        * it probably makes more sense to revisit this later in the course
+    * For simplicity, some of the examples will use less than ideal implementations
+
+
+* The same example described in :doc:`Topic3 </topic3.rst>` is used here
+    * A collection of ``Friend`` objects stored in an ``ContactList`` obejct
+
+
+Friend
+======
+
+* We will create a ``Friend`` class to keep track of a few details
+    * First Name
+    * Last Name
+    * Email
+
+* We will also want
+    * Constructor
+    * Some getters/setters
+
+* That's about it
+
+
+Header
+------
+
+* The first thing we want is a *header* file
+* Call this ``Friend.h``
+* This is where we typically declare all the member fields and methods the class will have
+* This file will also have all our ``#include``s too
+* Below is a complete ``Friend.h``
+
+
+.. code-block:: cpp
+    :linenos:
+
+    // Friend.h
+    #include <string>
+
+    class Friend {
+        private:
+            // Starting with underscore is convention for private variables
+            std::string _firstName;
+            std::string _lastName;
+            std::string _email;
+
+        public:
+            Friend();
+            Friend(std::string, std::string, std::string);
+            std::string getFirstName();
+            std::string getLastName();
+            std::string getEmail();
+            void setEmail(std::string);
```

Comment:
This is java standard, not cpp. Getters don't include the word `get` in cpp and return a reference, not a copy. For setters, you return a non `const` reference.

Reply:
For a property `prop`, you'd have a class field `_prop` and functions 
```cpp
const T& prop() const { return _prop; }
T& prop() { return _prop; }
```
For immutable fields, you wouldn't have the second one.

Reply:
C++14 introduced some other patterns with `std::move` and value references, but I don't know those as well.

Reply:
> This is java standard, not cpp. 

This was the method to the madness.

Reply:
The goal here is to inform those interested of cpp -- actually do that then. Don't teach some bastardization of the langauge. If there's idiomatic concepts to go into, then go into them.